Improve memory issues

RGoldsack opened the pull request:
Changed key creation step in line with https://github.com/jax-ml/jax/issues/17432 to fix memory leak when repeatedly calling estimators i.e. `bmi.estimators.HistogramEstimator()`. The updates are only to the estimator computation functions where the keys, x and y are directly used i.e. the function `estimate()` within class `HistogramEstimatorParams(BaseModel)`. 

Files for issue replication:
[bmi_estimate_mi.py](https://github.com/user-attachments/files/22326563/bmi_estimate_mi.py)
[memory_test.py](https://github.com/user-attachments/files/22326564/memory_test.py)

RGoldsack committed:
Update memory leak

Changed key creation step inline with https://github.com/jax-ml/jax/issues/17432 to fix memory leak when repeatedly calling estimators.

RGoldsack committed:
Removed .DS_Store files

No changes to functionality, just tidying up previously added .DS_Store files.

RGoldsack committed:
Remove remaining .DS_Store files

Same as last commit, trying to remove remaining .DS_Store files without more appearing

RGoldsack committed:
Create .DS_Store

RGoldsack committed:
Removing .DS_Store again

Finally seem to have removed all of the .DS_Store files

pawel-czyz commented on src/bmi/estimators/_histogram.py:
```diff
@@ -14,6 +14,7 @@
 from bmi.interface import BaseModel, IMutualInformationPointEstimator
 from bmi.utils import ProductSpace
 
+import gc # new
```
The comment `# new` will become redundant in a few months, so I'd suggest removing it from this PR.

pawel-czyz commented on src/bmi/estimators/neural/_basic_training.py:
```diff
@@ -72,10 +73,14 @@ def loss(critic, xs, ys):
     )
-    keys = jax.random.split(rng, max_n_steps)
-    for n_step, key in enumerate(keys, start=1):
+    # keys = jax.random.split(rng, max_n_steps) 
```
Very nice fix! Would you mind adding more in-text context, e.g., 

```python
# We don't use 
# keys = jax.random.split(rng, max_n_steps)
# because of memory leaks. See:
# https://github.com/jax-ml/jax/issues/17432
```

for future reference, so we don't forget the reason why to avoid this in the future?

pawel-czyz commented on src/bmi/estimators/neural/_critics.py:
```diff
@@ -54,6 +54,7 @@ def __init__(
         self.layers.append(eqx.nn.Linear(dims[-1], 1, key=key_final))
 
     def __call__(self, x: Point, y: Point) -> jax.Array:
+        # print(f"Critic - x shape {x.shape}, y shape {y.shape}")
```
I think this can be removed.

pawel-czyz commented on src/bmi/estimators/neural/_mine_estimator.py:
```diff
@@ -26,6 +26,7 @@
 from bmi.interface import BaseModel, EstimateResult, IMutualInformationPointEstimator
 from bmi.utils import ProductSpace
 
+import gc # new
```
The comment can be removed here as well.

pawel-czyz commented on src/bmi/estimators/neural/_mine_estimator.py:
```diff
@@ -245,8 +250,16 @@ def step(
+        xs_batch.delete()
+        ys_batch_paired.delete()
+        ys_batch_unpaired.delete()
+        # xs_test.delete()
```
Can `# xs_test.detele()` be removed as well?

pawel-czyz commented on src/bmi/estimators/neural/_mine_estimator.py:
```diff
@@ -245,8 +250,16 @@ def step(
+        ys_batch_unpaired.delete()
+        # xs_test.delete()
+        # ys_test_unpaired.delete()
+        del xs_batch, ys_batch_paired, ys_batch_unpaired #, xs_test, ys_test_unpaired
```
(Similarly here: the commented out variables should be removed).

pawel-czyz commented on src/bmi/estimators/neural/_mine_estimator.py:
```diff
@@ -245,8 +250,16 @@ def step(
+        del xs_batch, ys_batch_paired, ys_batch_unpaired #, xs_test, ys_test_unpaired
 
     training_log.finish()
+    jax.clear_caches()   # clears jit/compilation & staging caches
```
Very nice fix and explanation!

pawel-czyz reviewed (approved):
Hi,

thanks a lot for the improvements! I like them a lot :slightly_smiling_face:  
I only left some very minor comments, corresponding to some code clarity improvements. I think we can merge it as soon as they are addressed and the PR passes the GitHub CI checks :slightly_smiling_face: 
(I think that [running pre-commit as described here](https://cbg-ethz.github.io/bmi/contributing/) will resolve automatically all (or at least most) of the CI's complaints.) 

Thank you once again!

pawel-czyz committed:
Merge branch 'main' into Memory-Issues

RGoldsack committed:
Small updates for clarification

Removed '# new' from 'gc' imports and previous changes. Updated information regarding memory leaks in neural network basic training file and mentioned its change in MINE estimation. Removed deprecated for loops.
